### ISSUE TITLE
java.lang.ClassCastException in Jetty when using Spring Session

### ISSUE THREAD

**Issue by johanhaleby** [eclipse/jetty.project#3044]
I'm using Jetty 9.4.12 to which I deploy a Spring application using Spring Session with Redis. The application exposes a server-sent event (SSE) endpoint configured by Spring. But I seem to run into an issue when the javascript SSE client connection times out. I see the following log produced by Jetty:

```
[WARNING] java.lang.ClassCastException: org.springframework.session.web.http.SessionRepositoryFilter$SessionRepositoryRequestWrapper$HttpSessionWrapper 
    cannot be cast to org.eclipse.jetty.server.session.SessionHandler$SessionIf 
    while invoking onComplete listener org.eclipse.jetty.server.session.SessionHandler$SessionAsyncListener@67e50aa3
```

I.e. the [cast made in SessionAsyncListener](https://github.com/eclipse/jetty.project/blob/9d9189d47653a71bc2e4643f0f0013f55da69afd/jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionHandler.java#L366) fails. 

Is this a bug in Jetty? Is it correct to always assume that `session` is an instance of `SessionIf`? It doesn't seem to be the case for me at least.

For context I've also posted this as a question on [Stackoverflow](
https://stackoverflow.com/questions/53090592/java-lang-classcastexception-in-jetty-when-using-spring-session-with-redis-in-co).

**Comment by joakime** [comment]
SessionAsyncListener code ...

https://github.com/eclipse/jetty.project/blob/bc1e58b0f591311b5909029c13b0b5d620a18d49/jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionHandler.java#L170-L174

**Comment by joakime** [comment]
Do you have the complete stacktrace?

**Comment by johanhaleby** [comment]
Jetty doesn't just logs the last line and I've struggled to change this but I've not yet succeeded. However I've managed to debug and calling `exception.getStackTrace()` gives me the following:

```java
{java.lang.StackTraceElement[11]@17832} 
 0 = {java.lang.StackTraceElement@17833} "org.eclipse.jetty.server.session.SessionHandler.complete(SessionHandler.java:366)"
 1 = {java.lang.StackTraceElement@17834} "org.eclipse.jetty.server.session.SessionHandler$SessionAsyncListener.onComplete(SessionHandler.java:173)"
 2 = {java.lang.StackTraceElement@17835} "org.eclipse.jetty.server.HttpChannelState$4.run(HttpChannelState.java:889)"
 3 = {java.lang.StackTraceElement@17836} "org.eclipse.jetty.server.handler.ContextHandler.handle(ContextHandler.java:1431)"
 4 = {java.lang.StackTraceElement@17837} "org.eclipse.jetty.server.HttpChannelState.runInContext(HttpChannelState.java:1116)"
 5 = {java.lang.StackTraceElement@17838} "org.eclipse.jetty.server.HttpChannelState.onComplete(HttpChannelState.java:905)"
 6 = {java.lang.StackTraceElement@17839} "org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:508)"
 7 = {java.lang.StackTraceElement@17840} "org.eclipse.jetty.server.HttpChannel.run(HttpChannel.java:305)"
 8 = {java.lang.StackTraceElement@17841} "org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:765)"
 9 = {java.lang.StackTraceElement@17842} "org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:683)"
 10 = {java.lang.StackTraceElement@17843} "java.lang.Thread.run(Thread.java:748)"
```

**Comment by gregw** [comment]
I think it is kind of strange and fragile to wrap the session object.... But obviously it is done.  We can use our Request.getBaseRequest method to navigate to the unwrapped request and get the unwrapped session from it.

**Comment by johanhaleby** [comment]
@gregw 

Do you mean that instead of doing this?

```java
complete(((HttpServletRequest)event.getAsyncContext().getRequest()).getSession(false));
```

we could do:

```java
HttpSession session = ((HttpServletRequest)event.getAsyncContext().getRequest()).getSession(false)
complete(Request.getBaseRequest(session));
```

Or do you mean instead of the [cast inside SessionHandler](https://github.com/eclipse/jetty.project/blob/9d9189d47653a71bc2e4643f0f0013f55da69afd/jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionHandler.java#L366)?

Btw, is this warning something that we could ignore for now? Or will there resource leaks or something similar that will accumulate over time?

**Comment by gregw** [comment]
I was thinking:
```
complete(Request.getBaseRequest(event.getAsyncContext().getRequest()).getSession(false));
```

Depending on what session store you are using, the warnings are probably not good and could mean that sessions are not being passivated correctly.

We have a release due this week... so @janbartel can you fix (assuming the fix is OK?) this before that?

**Comment by janbartel** [comment]
Fixed for 9.4.13.

**Comment by joakime** [comment]
Jetty version `9.4.13.v20181111` is now available on Maven Central.
